### PR TITLE
[RPA-1669] Add credential selection for specific action on bot

### PR DESCRIPTION
**Pull request by marcin-depcik**
## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/docker-compose.yml [hunk 1]
```diff
@@ -57,6 +57,7 @@ services:
       - MAIL_PORT=
       - MAIL_USERNAME=
       - MAIL_PASSWORD=
+      - ENCRYPTION_KEY=O3OTmXJpIIyUb3ByT4t65xblgsxfDGQu
```
Comment (star-in-the-jar) [inline runbotics/docker-compose.yml:60]: not a real one, right?

File: runbotics/runbotics-desktop/src/action/beeoffice/bee-office.action-handler.ts [hunk 2]
```diff
@@ -297,10 +298,15 @@ export default class BeeOfficeActionHandler extends StatelessActionHandler {
     }
 
     run(request: BeeOfficeActionRequest) {
-        const credential: BeeOfficeCredential = { // @todo method for matching credentialId (action template) with decrypted credential values
-            ...(this.serverConfigService.beeAuth), // @todo skip grant_type as it's being hardcoded later
+        const passwordManagerCredential =
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/beeoffice/bee-office.action-handler.ts]: not the best fit for a variable name

i suggest matchedCredential

File: runbotics/runbotics-desktop/src/action/google/google.action-handler.ts [hunk 3]
```diff
@@ -141,35 +143,41 @@ export default class GoogleActionHandler extends StatelessActionHandler {
...
     run(request: GoogleActionRequest) {
+        const passwordManagerCredential =
+            credentialAttributesMapper<GoogleCredential>(request.credentials);
+
+        // @todo After completion of password manager switch fully to passwordManagerCredential
+        const credential: GoogleCredential = passwordManagerCredential ?? {
+            email: this.serverConfigService.googleAuth.serviceAccountEmail,
+            key: this.serverConfigService.googleAuth.privateKey,
+        };
+
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/google/google.action-handler.ts]: what a guy
did my job 🥇

File: runbotics/runbotics-desktop/src/action/mail/mail.action-handler.ts [hunk 4]
```diff
@@ -73,11 +74,15 @@ export default class MailActionHandler extends StatelessActionHandler {
     }
 
     run(request: MailActionRequest) {
-        const mailCredential: MailCredential = { // @todo replace with method for matching credentialId/templateName with decrypted credentials
+        const passwordManagerCredential =
+            credentialAttributesMapper<MailCredential>(request.credentials);
+
+        // @todo After completion of password manager switch fully to passwordManagerCredential
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/mail/mail.action-handler.ts]: good You've left comments

File: runbotics/runbotics-desktop/src/action/mail/mail.action-handler.ts [hunk 5]
```diff
@@ -73,11 +74,15 @@ export default class MailActionHandler extends StatelessActionHandler {
...
+        const mailCredential: MailCredential = passwordManagerCredential ?? {
             mailUsername: this.serverConfigService.mailUsername,
             mailPassword: this.serverConfigService.mailPassword,
             mailHost: this.serverConfigService.mailHost,
-            mailPort: Number(this.serverConfigService.mailPort),
+            mailPort: this.serverConfigService.mailPort,
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/mail/mail.action-handler.ts:85]: I'm afraid this is not going to work - env values are string, and for mailPort number is expected

File: runbotics/runbotics-desktop/src/action/mail/mail.types.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 export interface MailCredential {
     mailHost: string;
-    mailPort: number;
+    mailPort: string | number;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/action/mail/mail.types.ts:3]: we don't want string

Comment (star-in-the-jar) [reply runbotics/runbotics-desktop/src/action/mail/mail.types.ts:3]: as far as I know

File: runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts [hunk 7]
```diff
@@ -685,4 +692,23 @@ export class RuntimeService implements OnApplicationBootstrap, OnModuleDestroy {
         content.isSequenceFlow &&
         content.sourceId.includes('Gateway_')
     );
+
+    private determineCredentialsForAction(
+        credentialId: string | undefined,
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts]: or null. And there should be type CredentialId used. define it if it's not there yet

File: runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts [hunk 8]
```diff
@@ -685,4 +692,23 @@ export class RuntimeService implements OnApplicationBootstrap, OnModuleDestroy {
...
+        credentialTemplateName: string,
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts]: this should be enum ActionCredentialType

File: runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts [hunk 9]
```diff
@@ -685,4 +692,23 @@ export class RuntimeService implements OnApplicationBootstrap, OnModuleDestroy {
...
+        credentials: DecryptedCredential[]
+    ) {
+        if (!credentials) {
+            return [];
+        }
+        if (credentialId) {
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.service.ts]: else if

File: runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.types.ts [hunk 10]
```diff
@@ -41,6 +41,7 @@ export interface RunBoticsExecutionEnvironment extends BpmnEngineExecutionEnviro
     runbotic?: {
         disabled: boolean;
         processOutput: boolean;
+        credentialType?: string;
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/core/bpm/runtime/runtime.types.ts]: enum type

File: runbotics/runbotics-desktop/src/mailer/mailer.service.ts [hunk 11]
```diff
@@ -8,7 +8,7 @@ export class MailService {
     async sendMail({ to, from, subject, html }: Mail.Options, auth: MailCredential) {
         const transporter = nodemailer.createTransport({
             host: auth.mailHost,
-            port: auth.mailPort,
+            port: Number(auth.mailPort),
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/mailer/mailer.service.ts]: honor returned

File: runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
+import { DecryptedCredential } from 'runbotics-common';
+
+export const credentialAttributesMapper = <R = Record<string, any>>(
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts:9]: very interesting solution

File: runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    const { attributes } = credentials[0];
+    if (!attributes.length) {
+        throw new Error('Credential must have at least one attribute');
+    }
+
+    const mappedCredential = attributes
+        .reduce((acc, { name, value }) => {
+            acc[name] = value;
+            return acc;
+        }, {} as R);
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts:26]: have you tried {} satisfies R

File: runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import { DecryptedCredential } from 'runbotics-common';
+
```
Comment (star-in-the-jar) [inline runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts:2]: could use little documentation for future generations. in this convention
 
image

File: runbotics/runbotics-scheduler/src/migrations/1727274592337-GoogleCredentialTemplate.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
+import { CredentialTemplateAttribute } from '#/scheduler-database/credential-template-attribute/credential-template-attribute.entity';
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/migrations/1727274592337-GoogleCredentialTemplate.ts:1]: owe you

**Review by star-in-the-jar — APPROVED** [review]

File: runbotics/docker-compose.yml (hunk 1 above)
Comment (marcin-depcik) [reply runbotics/docker-compose.yml:60]: tell me, what is real these days??

File: runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts (hunk 13 above)
Comment (marcin-depcik) [reply runbotics/runbotics-desktop/src/utils/credentialAttributesMapper.ts:26]: can't do

**Pull request merged**